crates/fuel-core/src/graphql_api/indexation/coins_to_spend.rs

```diff
@@ -0,0 +1,739 @@
+        collection::vec,
+        prelude::*,
+    };
+
+    use crate::{
+        database::{
+            database_description::off_chain::OffChain,
+            Database,
+        },
+        graphql_api::{
+            indexation::{
+                coins_to_spend::{
+                    update,
+                    RETRYABLE_BYTE,
+                },
+                error::IndexationError,
+                test_utils::{
+                    make_coin,
+                    make_nonretryable_message,
+                    make_retryable_message,
+                },
+            },
+            storage::coins::{
+                CoinsToSpendIndex,
+                CoinsToSpendIndexKey,
+            },
+        },
+    };
+
+    use super::NON_RETRYABLE_BYTE;
+
+    fn assert_index_entries(
+        db: &Database<OffChain>,
+        expected_entries: &[(Address, AssetId, [u8; 1], u64)],
+    ) {
+        let actual_entries: Vec<_> = db
+            .entries::<CoinsToSpendIndex>(None, IterDirection::Forward)
+            .map(|entry| entry.expect("should read entries"))
+            .map(|entry| {
+                (
+                    entry.key.owner(),
+                    entry.key.asset_id(),
+                    [entry.key.retryable_flag()],
+                    entry.key.amount(),
+                )
+            })
+            .collect();
+
+        assert_eq!(expected_entries, actual_entries.as_slice());
+    }
+
+    #[test]
+    fn coins_to_spend_indexation_enabled_flag_is_respected() {
+        use tempfile::TempDir;
+        let tmp_dir = TempDir::new().unwrap();
+        let mut db: Database<OffChain> =
+            Database::open_rocksdb(tmp_dir.path(), None, Default::default(), 512)
+                .unwrap();
+        let mut tx = db.write_transaction();
+
+        // Given
+        const COINS_TO_SPEND_INDEX_IS_DISABLED: bool = false;
+        let base_asset_id = AssetId::from([0; 32]);
+
+        let owner_1 = Address::from([1; 32]);
+        let owner_2 = Address::from([2; 32]);
+
+        let asset_id_1 = AssetId::from([11; 32]);
+        let asset_id_2 = AssetId::from([12; 32]);
+
+        let coin_1 = make_coin(&owner_1, &asset_id_1, 100);
+        let coin_2 = make_coin(&owner_1, &asset_id_2, 200);
+        let message_1 = make_retryable_message(&owner_1, 300);
+        let message_2 = make_nonretryable_message(&owner_2, 400);
+
+        // TODO[RC]: No clone() required for coins? Double check the types used,
+        // maybe we want `MessageCoin` (which is Copy) for messages?
+        // 1) Currently we use the same types as embedded in the executor `Event`.
+        // 2) `MessageCoin` will refuse to construct itself from a `Message` if the data is empty
+        //      impl TryFrom<Message> for MessageCoin { ... if !data.is_empty() ... }
+        // Actually it shouldn't matter from the indexation perspective, as we just need
+        // to read data from the type and don't care about which data type we took it from.
```

Comment:
This seems resolved, should we delete it?
```suggestion
```

Reply:
No, actually it's not solved. I mean, the code is most likely correct, but what caught my attention is that in the code which handles coins to spend:
1. coins are represented by [entities::coins::Coin](https://github.com/FuelLabs/fuel-core/blob/master/crates/types/src/entities/coins/coin.rs#L23) , which is `Copy`
2. messages are represented by [entities::relayer::Message](https://github.com/FuelLabs/fuel-core/blob/master/crates/types/src/entities/relayer/message.rs#L34) , which is not `Copy`

For a moment I thought that we should use [`MessageCoin`](https://github.com/FuelLabs/fuel-core/blob/master/crates/types/src/entities/coins/message_coin.rs#L15) for messages, but we cannot construct retryable message coin due to [this](https://github.com/FuelLabs/fuel-core/blob/master/crates/types/src/entities.rs#L27-L31).

So, it's not an issue with `Copy` vs `non-Copy`, but potentially with incorrect type being used (still unlikely).

cc: @xgreenx - I'd like to double check this with you. If the current code is ok, I'll just remove `TODO`.